### PR TITLE
Add size-optimized FFI functions for Locale and DataProvider

### PR DESCRIPTION
**Pull request by sffc**
This PR copies the fixeddecimal example to use small-code functions.

*optim5.elf* is 56 KB after this change (60 KB if I keep the additional coverage tests).

### CODE REVIEW

File: ffi/capi/examples/fixeddecimal/test.c [hunk 1]
```diff
@@ -7,18 +7,18 @@
 #include <stdio.h>
 
 int main() {
-    ICU4XLocale* locale = ICU4XLocale_create("bn", 2);
-    ICU4XCreateDataProviderResult result = ICU4XDataProvider_create_static();
+    ICU4XLocale* locale = ICU4XLocale_create_bn();
+    ICU4XCreateFixedDecimalFormatDataProviderResult result = ICU4XFixedDecimalFormatDataProvider_create_static();
```
Comment (Manishearth) [inline ffi/capi/examples/fixeddecimal/test.c:11]: thought: Given these changes should we split this example into an entirely separate one? That way we can have a "normal" FDF example as a real example, and then the size optimized one can be as weird as it likes

**Review by Manishearth — DISMISSED** [review]

**Comment by sffc** [comment]
The CI error is

```
  = note: /usr/bin/ld: error: LLVM gold plugin has failed to create LTO module: Invalid record
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I cannot reproduce this locally on the specified toolchain version (nightly-2021-02-28), although I can reproduce it on other toolchain versions like nightly-2021-08-18.

I suspect the issue is most likely a different version of `lld` on CI than I have on my local computer.  According to https://github.com/rust-lang/rust/issues/60059#issuecomment-823342782, it is necessary to use a specific version of `lld` found in Rust's fork of LLVM.

Should I:

1. Keep trying to get a build running in CI
2. Remove the CI for now and work on it in a follow-up

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
```
  = note: /usr/bin/ld: error: LLVM gold plugin has failed to create LTO module: Invalid record
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

in test-capi-tiny

**Comment by sffc** [comment]
> ```
>   = note: /usr/bin/ld: error: LLVM gold plugin has failed to create LTO module: Invalid record
>           clang: error: linker command failed with exit code 1 (use -v to see invocation)
> ```
> 
> in test-capi-tiny

Thoughts on the question I posed in my previous post?

**Comment by Manishearth** [comment]
I'd remove the CI for now and document the specific steps.

**Review by Manishearth — APPROVED** [review]

**Pull request merged**